src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs

```diff
@@ -0,0 +1,373 @@
+        {
+            var list = new List<string>();
+            var values = headers[headerName];
+            if (values != null)
+            {
+                list.Add(values.TrimStart(null));
+            }
+
+            return list;
+        }
+
+        /// <summary>
+        /// The extract query key values.
+        /// </summary>
+        /// <param name="address">
+        /// The address.
+        /// </param>
+        /// <returns>
+        /// The <see>
+        ///         <cref>Dictionary</cref>
+        ///     </see>
+        ///     .
+        /// </returns>
+        private static Dictionary<string, HashSet<string>> ExtractQueryKeyValues(Uri address)
+        {
+            var values = new Dictionary<string, HashSet<string>>();
+            var newreg = new Regex(@"\?(\w+)\=([\w|\=]+)|\&(\w+)\=([\w|\=]+)");
+            var matches = newreg.Matches(address.Query);
+            foreach (Match match in matches)
+            {
+                string key, value;
+                if (!string.IsNullOrEmpty(match.Groups[1].Value))
+                {
+                    key = match.Groups[1].Value;
+                    value = match.Groups[2].Value;
+                }
+                else
+                {
+                    key = match.Groups[3].Value;
+                    value = match.Groups[4].Value;
+                }
+
+                if (!values.ContainsKey(key))
+                {
+                    var newSet = new HashSet<string> { value };
+                    values.Add(key, newSet);
+                }
+                else
+                {
+                    HashSet<string> setOfValues = null;
+                    values.TryGetValue(key, out setOfValues);
+                    if (setOfValues != null)
+                    {
+                        setOfValues.Add(value);
+                        values.Remove(key);
+                        values.Add(key, setOfValues);
+                    }
+                }
+            }
+
+            return values;
+        }
+
+        /// <summary>
+        /// The get comma separated list.
+        /// </summary>
+        /// <param name="queryKeyValues">
+        /// The query key values.
+        /// </param>
+        /// <returns>
+        /// The <see>
+        ///         <cref>Dictionary</cref>
+        ///     </see>
+        ///     .
+        /// </returns>
+        private static Dictionary<string, HashSet<string>> GetCommaSeparatedList(
+            Dictionary<string, HashSet<string>> queryKeyValues)
+        {
+            var dictionary = new Dictionary<string, HashSet<string>>();
+            
+            foreach (var queryKeys in queryKeyValues.Keys)
+            {
+                HashSet<string> setOfValues = null;
+                queryKeyValues.TryGetValue(queryKeys, out setOfValues);
+                var listArray = setOfValues.ToArray();
+                var list = new List<string>();
+                list.AddRange(listArray);
```

Comment:
List.AddRange should be able to take the HashSet... why is there an explicit intermediate array?